obsercer_pattern.java

```diff
@@ -0,0 +1,75 @@
+import java.util.ArrayList;
+import java.util.List;
+
+// Subject interface
+interface Subject {
+    void addObserver(Observer observer);
+    void removeObserver(Observer observer);
+    void notifyObservers();
+}
+
+// Concrete Subject
+class ConcreteSubject implements Subject {
+    private List<Observer> observers = new ArrayList<>();
+    private int state;
+
+    public int getState() {
+        return state;
+    }
+
+    public void setState(int state) {
+        this.state = state;
+        notifyObservers();
+    }
+
+    @Override
+    public void addObserver(Observer observer) {
+        observers.add(observer);
+    }
+
+    @Override
+    public void removeObserver(Observer observer) {
+        observers.remove(observer);
+    }
+
+    @Override
+    public void notifyObservers() {
+        for (Observer observer : observers) {
+            observer.update();
+        }
+    }
+}
+
+// Observer interface
+interface Observer {
+    void update();
+}
+
+// Concrete Observer
+class ConcreteObserver implements Observer {
+    private ConcreteSubject subject;
+
+    public ConcreteObserver(ConcreteSubject subject) {
+        this.subject = subject;
+        this.subject.addObserver(this);
+    }
+
+    @Override
+    public void update() {
+        System.out.println("Observer updated. New state: " + subject.getState());
+    }
+}
+
+public class Main {
+    public static void main(String[] args) {
+        ConcreteSubject subject = new ConcreteSubject();
+        ConcreteObserver observer1 = new ConcreteObserver(subject);
```

Comment:
OH, I think I understand, you pass in the subject that you're wanting to observer on the observer creation?

if you wanted to have an observer observe multiple subjects, how would you go about that?

I was going to suggest having conditional statements in your subscriber classes to see if an observer is already subscribed or not to a subject, but that would most likely be redundant?